File: mstparser/Alphabet.java
/* Copyright (C) 2002 Univ. of Massachusetts Amherst, Computer Science Dept.
   This file is part of "MALLET" (MAchine Learning for LanguagE Toolkit).
   http://www.cs.umass.edu/~mccallum/mallet
   This software is provided under the terms of the Common Public License,
   version 1.0, as published by http://www.opensource.org.  For further
   information, see the file `LICENSE' included with this distribution. */




/** 
    @author Andrew McCallum <a href="mailto:mccallum@cs.umass.edu">mccallum@cs.umass.edu</a>
*/

package mstparser;

import java.util.ArrayList;
import java.io.*;
import java.util.Iterator;

public class Alphabet implements Serializable
{
    gnu.trove.TObjectIntHashMap map;
    int numEntries;
    boolean growthStopped = false;

    public Alphabet (int capacity)
    {
	this.map = new gnu.trove.TObjectIntHashMap (capacity);
	numEntries = 0;
    }

    public Alphabet ()
    {
	this (10000);
    }

	
    /** Return -1 if entry isn't present. */
    public int lookupIndex (Object entry, boolean addIfNotPresent)
    {
	if (entry == null)
	    throw new IllegalArgumentException ("Can't lookup \"null\" in an Alphabet.");
	int ret = map.get(entry);
	if (ret == -1 && !growthStopped && addIfNotPresent) {
	    ret = numEntries;
	    map.put (entry, ret);
	    numEntries++;
	}
	return ret;
    }

    public int lookupIndex (Object entry)
    {
	return lookupIndex (entry, true);
    }
	
    public Object[] toArray () {
	return map.keys();
    }

    public boolean contains (Object entry)
    {
	return map.contains (entry);
    }

    public int size ()
    {
	return numEntries;
    }

    public void stopGrowth ()
    {
	growthStopped = true;
    }

    public void allowGrowth ()
    {
	growthStopped = false;
    }

    public boolean growthStopped ()
    {
	return growthStopped;
    }


    // Serialization 
		
    private static final long serialVersionUID = 1;
    private static final int CURRENT_SERIAL_VERSION = 0;

    private void writeObject (ObjectOutputStream out) throws IOException {
	out.writeInt (CURRENT_SERIAL_VERSION);
	out.writeInt (numEntries);
	/*
	Object[] keys = map.keys();
	for(int i = 0; i < keys.length; i++) {
	    out.writeObject(keys[i]); out.writeInt(map.get(keys[i]));
	}
	*/
	out.writeObject(map);
	out.writeBoolean (growthStopped);
    }

    private void readObject (ObjectInputStream in) throws IOException, ClassNotFoundException {
	int version = in.readInt ();
	numEntries = in.readInt();
	/*
	map =  new gnu.trove.TObjectIntHashMap(numEntries);
	for(int i = 0; i < keys.length; i++) {
	    map.put(in.readObject(),in.readInt());
	}
	*/
	map = (gnu.trove.TObjectIntHashMap)in.readObject();
	growthStopped = in.readBoolean();
    }
	
}

File: src/main/scala/mstparser/Alphabet.scala
package mstparser

import gnu.trove.map.hash.TObjectIntHashMap

class Alphabet[A](capacity: Int) extends Serializable {
  def this() = this(10000)

  private[this] val map = new TObjectIntHashMap[A](capacity, 0.75F, -1)
  private[this] var growing = true
  private[this] val _values = scala.collection.mutable.ArrayBuffer.empty[A]

  def lookupIndex(entry: A) = this.map.get(entry) match {
    case -1 if this.growing =>
      this._values += entry
      this.map.put(entry, this.map.size)
      this.map.size - 1
    case i => i
  }

  def values: IndexedSeq[A] = this._values
  def size = this.values.size

  def stopGrowth() {
    this.map.compact()
    this.growing = false
  }
}

/*class Alphabet[A](capacity: Int) extends Serializable {
  def this() = this(10000)

  @transient private[this] val map = new scala.collection.mutable.HashMap[A, Int].withDefaultValue(-1)
  private[this] var growing = true
  private[this] val _values = scala.collection.mutable.ArrayBuffer.empty[A]

  def lookupIndex(entry: A) = if (this.growing) {
    //this._values += entry
    this.map.getOrElseUpdate(entry, { this._values += entry; this.map.size })
  } else this.map(entry)
    /*this.map.get(entry) match {
    case -1 if this.growing =>
      this._values += entry
      this.map.put(entry, this.map.size)
      this.map.size - 1
    case i => i
  }*/

  def values: IndexedSeq[A] = this._values
  def size = this.values.size

  def stopGrowth() {
    //this.map.compact()
    this.growing = false
  }
 
  private def writeObject(out: java.io.ObjectOutputStream) {
    out.writeObject(this._values)
    out.writeObject(this.growing)
  }

  private def readObject(in: java.io.ObjectInputStream) {
  }
}*/

